### PR TITLE
Add preliminary macro infrastructure from gestalt

### PR DESCRIPTION
**Pull request by olafurpg**
- sets up bintray publishing so we can immediately use these changes from scalacenter/macros
- bring in all the changes from @liufengyun from the gestalt fork of Dotty
- revert and adapt a few changes from the gestalt fork

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/StdNames.scala [hunk 1]
```diff
@@ -459,6 +459,7 @@ object StdNames {
     val materializeWeakTypeTag: N = "materializeWeakTypeTag"
     val materializeTypeTag: N   = "materializeTypeTag"
     val mirror : N              = "mirror"
+    val meta : N                = "meta"
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/StdNames.scala]: We need to change to `macro`? No, we can just remove this line, as `macro` already exists.

File: compiler/src/dotty/tools/dotc/macros/Transform.scala [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+import reporting.diagnostic.NoExplanation
+import reporting.diagnostic.Message._
+import config.Printers.{ macros => debug }
+
+/** Transform macros definitions
+ *
+ *  Macro definition is transformed from:
+ *
+ *    class macros {
+ *      def f[T](a: A)(b: B): C = meta {
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/macros/Transform.scala]: change `meta` to `macro`

File: compiler/src/dotty/tools/dotc/macros/Transform.scala [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    }
+
+    val bodyNew = macrosNew ++ tmpl.body.diff(macros)
+    cpy.Template(tmpl)(body = bodyNew)
+  }
+
+  def getMacros(tmpl: Template)(implicit ctx: Context): List[DefDef] = tmpl.body.filter {
+    case mdef : DefDef =>
+      mdef.rhs match {
+        case Apply(Ident(nme.meta), _) => true
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/macros/Transform.scala]: change `nme.meta` to `nme.macro`.

File: compiler/src/dotty/tools/dotc/macros/Transform.scala [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        case _ => false
+      }
+    case _ => false
+  }.asInstanceOf[List[DefDef]]
+
+
+  /** Create macro implementation method
+   *
+   *     def f[T](a: A)(b: B): C = meta { body }
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/macros/Transform.scala]: change `meta` to `macro`.

File: compiler/src/dotty/tools/dotc/macros/macros.scala [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+   *
+   *  @param tree  the tree that may contain macro definition, the tree may be a thicket
+   *  @returns     the transformed tree
+   *
+   *  @note The returned tree NEEDs desugaring
+   *
+   *  Macro definition is transformed from:
+   *
+   *    class macros {
+   *      def f[T](a: A)(b: B): C = meta {
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/macros/macros.scala]: change `meta` to `macro`.

File: compiler/src/dotty/tools/dotc/macros/macros.scala [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+   *    }
+   *
+   *  to:
+   *
+   *    class main {
+   *      <macro> def f[T](a: A)(b: B): C = ???
+   *    }
+   *
+   *    object main$inline {
+   *      @static def f(prefix: scala.meta.Term)(T: scala.meta.Type)(a: scala.meta.Term)(b: scala.meta.Term): scala.meta.Tree = body
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/macros/macros.scala]: change `scala.meta` to `scala.macros`

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 7]
```diff
@@ -1858,7 +1858,8 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
       if (tree.isDef) interpolateUndetVars(tree, tree.symbol)
       else if (!tree.tpe.widen.isInstanceOf[LambdaType]) interpolateUndetVars(tree, NoSymbol)
       tree.overwriteType(tree.tpe.simplified)
-      adaptInterpolated(tree, pt)
+      val tpdTree = adaptInterpolated(tree, pt)
+      macros.expandDefMacro(tpdTree, pt, this)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:1862]: It's nice that we only have one line change to the typer.

**Review by liufengyun — APPROVED** [review]
Some minor changes, otherwise LGTM.

File: compiler/src/dotty/tools/dotc/core/StdNames.scala (hunk 1 above)
Comment (olafurpg) [reply compiler/src/dotty/tools/dotc/core/StdNames.scala]: Done.

File: compiler/src/dotty/tools/dotc/macros/Transform.scala (hunk 2 above)
Comment (olafurpg) [reply compiler/src/dotty/tools/dotc/macros/Transform.scala]: Done.

File: compiler/src/dotty/tools/dotc/macros/Transform.scala (hunk 3 above)
Comment (olafurpg) [reply compiler/src/dotty/tools/dotc/macros/Transform.scala]: Done.

File: compiler/src/dotty/tools/dotc/macros/Transform.scala (hunk 4 above)
Comment (olafurpg) [reply compiler/src/dotty/tools/dotc/macros/Transform.scala]: Done.

File: compiler/src/dotty/tools/dotc/macros/macros.scala (hunk 5 above)
Comment (olafurpg) [reply compiler/src/dotty/tools/dotc/macros/macros.scala]: Done.

File: compiler/src/dotty/tools/dotc/macros/macros.scala (hunk 6 above)
Comment (olafurpg) [reply compiler/src/dotty/tools/dotc/macros/macros.scala]: Done.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 7 above)
Comment (olafurpg) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala:1862]: Pretty amazing 😄

**Review by olafurpg — COMMENTED** [review]
Thank you for the review @liufengyun !

**Pull request merged**